tests/cpp/test_frontend.cc

```diff
@@ -5,16 +5,498 @@
+  auto input_stream = rapidjson::MemoryStream(json_str.c_str(),
+                                              json_str.size());
+  std::shared_ptr<MockDelegator> delegator =
+    std::make_shared<MockDelegator>();
+
+  Model output;
+  details::GradientBoosterHandler wrapped_handler {delegator, output};
+  MockObjectStarter handler {delegator, wrapped_handler};
+
+  rapidjson::Reader reader;
+  auto result = reader.Parse(input_stream, handler);
+
+  ASSERT_TRUE(result);
+}
+
+/******************************************************************************
+ * ObjectiveHandler
+ * ***************************************************************************/
+TEST(ObjectiveHandlerSuite, ObjectiveHandler) {
+  std::string json_str = "{\"name\": \"multi:softmax\"}";
+  auto input_stream = rapidjson::MemoryStream(json_str.c_str(),
+                                              json_str.size());
+  std::shared_ptr<MockDelegator> delegator =
+    std::make_shared<MockDelegator>();
+
+  std::string output;
+  details::ObjectiveHandler wrapped_handler {delegator, output};
+  MockObjectStarter handler {delegator, wrapped_handler};
+
+  rapidjson::Reader reader;
+  auto result = reader.Parse(input_stream, handler);
+
+  ASSERT_TRUE(result);
+
+  ASSERT_EQ(output, "multi:softmax");
+}
+
+/******************************************************************************
+ * LearnerParamHandler
+ * ***************************************************************************/
+TEST(LearnerParamHandlerSuite, LearnerParamHandler) {
+  std::string json_str =
+    "{\"base_score\": \"0.5\", \"num_class\": \"5\", \"num_feature\": \"12\"}";
+  auto input_stream = rapidjson::MemoryStream(json_str.c_str(),
+                                              json_str.size());
+  std::shared_ptr<MockDelegator> delegator =
+    std::make_shared<MockDelegator>();
+
+  Model output;
+  details::LearnerParamHandler wrapped_handler {delegator, output};
+  MockObjectStarter handler {delegator, wrapped_handler};
+
+  rapidjson::Reader reader;
+  auto result = reader.Parse(input_stream, handler);
+
+  ASSERT_TRUE(result);
+
+  ASSERT_FLOAT_EQ(output.param.global_bias, 0.5);
+  ASSERT_EQ(output.num_output_group, 5);
+  ASSERT_EQ(output.num_feature, 12);
+}
+
+/******************************************************************************
+ * XGBoostModelHandler
+ * ***************************************************************************/
+TEST(XGBoostModelHandlerSuite, XGBoostModelHandler) {
+  std::string json_str = "{\"version\": []}";
+  auto input_stream = rapidjson::MemoryStream(json_str.c_str(),
+                                              json_str.size());
+  std::shared_ptr<MockDelegator> delegator =
+    std::make_shared<MockDelegator>();
+
+  Model output;
+  details::LearnerParamHandler wrapped_handler {delegator, output};
+  MockObjectStarter handler {delegator, wrapped_handler};
+
+  rapidjson::Reader reader;
+  auto result = reader.Parse(input_stream, handler);
+}
```

Comment:
This test does not have any assertion. Is it sufficient that the test runs without crashing?

Reply:
Yes, that's right. Because this is at such a high level in the tree, there are very few non-trivial checks that one could make for correctness here without invoking a larger stack of delegated handlers and thereby turning this into an integration test rather than a unit test. Since the Python test exercises the full stack, this unit test is designed solely to ensure that the C++ code does not immediately die at this level of parsing.

That being said, there's no reason to store the result here (unused variable). I'll pull that out.

Reply:
Just added an explanatory comment to this and the other test without assertions. Also pulled out the unused variable from both.